File: trees_and_graphs/reachable_nodes_with_restrictions.java
public int reachableNodesBFS(int n, int[][] edges, int[] restricted)
{
    // Store all edges in neighbors
    Map<Integer, List<Integer>> neighbors = new HashMap<>();
    for (int[] edge : edges)
    {
        int a = edge[0];
        int b = edge[1];

        neighbors.computeIfAbsent(a, value -> new ArrayList<Integer>()).add(b);
        neighbors.computeIfAbsent(b, value -> new ArrayList<Integer>()).add(a);
    }

    // Mark the nodes in restricted as visited
    Set<Integer> seen = new HashSet<>();
    for (int node : restricted)
    {
        seen.add(node);
    }

    // Store all the nodes to be visited in queue
    int ans = 0;
    seen.add(0);

    Queue<Integer> queue = new LinkedList<>(Arrays.asList(0));

    while (!queue.isEmpty())
    {
        int currNode = queue.poll();
        ans++;

        for (int nextNode : neighbors.get(currNode))
        {
            if (!seen.contains(nextNode))
            {
                seen.add(nextNode);
                queue.offer(nextNode);
            }
        }
    }

    return ans;
}

class Solution
{
    private int ans = 0;

    private void dfsRecursive(int                         currNode,
                              Map<Integer, List<Integer>> neighbors,
                              Set<Integer>                seen)
    {
        // Mark currNode as visited and increment ans by 1
        ans++;
        seen.add(currNode);

        // Go for unvisited neighbors of currNode
        for (int nextNode : neighbors.get(currNode))
        {
            if (!seen.contains(nextNode))
            {
                dfsRecursive(nextNode, neighbors, seen);
            }
        }
    }

    public int reachableNodesRecursive(int n, int[][] edges, int[] restricted)
    {
        // Store all edges according to nodes in neighbors
        Map<Integer, List<Integer>> neighbors = new HashMap<>();

        for (int[] edge : edges)
        {
            int a = edge[0];
            int b = edge[1];

            neighbors.computeIfAbsent(a, value -> new ArrayList<Integer>()).add(b);
            neighbors.computeIfAbsent(b, value -> new ArrayList<Integer>()).add(a);
        }

        // Mark nodes in restricted as visited
        Set<Integer> seen = new HashSet<>();
        for (int node : restricted)
        {
            seen.add(node);
        }

        dfsRecursive(0, neighbors, seen);
        return ans;
    }
}

public int reachableNodesIterative(int n, int[][] edges, int[] restricted)
{
    // Store all edges according to nodes in neighbors
    Map<Integer, List<Integer>> neighbors = new HashMap<>();
    for (int[] edge : edges)
    {
        int a = edge[0];
        int b = edge[1];

        neighbors.computeIfAbsent(a, value -> new ArrayList<Integer>()).add(b);
        neighbors.computeIfAbsent(b, value -> new ArrayList<Integer>()).add(a);
    }

    // Mark the nodes in restricted as visited
    Set<Integer> seen = new HashSet<>();
    for (int node : restricted)
    {
        seen.add(node);
    }

    // Use stack to store all nodes to be visited
    int ans = 0;
    seen.add(0);
    Stack<Integer> stack = new Stack<>();
    stack.add(0);

    while (!stack.isEmpty())
    {
        int currNode = stack.pop();
        ans++;

        // Add all unvisited neighbors of current node to stack
        // and mark it as visited
        for (int nextNode : neighbors.get(curNode))
        {
            if (!seen.contains(nextNode))
            {
                seen.add(nextNode);
                stack.add(nextNode);
            }
        }
    }

    return ans;
}

class UnionFind
{
    private int[] root;
    private int[] rank;

    public UnionFind(int n)
    {
        this.root = new int[n];
        this.rank = new int[n];

        for (int i = 0; i < n; ++i)
        {
            this.root[i] = i;
            this.rank[i] = 1;
        }
    }

    public int find(int x)
    {
        if (this.root[x] != x)
        {
            this.root[x] = find(this.root[x]);
        }

        return this.root[x];
    }

    public void union(int x, int y)
    {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY)
        {
            if (this.rank[rootX] > this.rank[rootY])
            {
                int tmp = rootX;
                rootX = rootY;
                rootY = tmp;
            }
            this.root[rootX] = rootY;
            this.rank[rootY] += this.rank[rootX];
        }
    }

    public int getSize(int x)
    {
        return this.rank[find(x)];
    }
}

public int reachableNodes(int n, int[][] edges, int[] restricted)
{
    UnionFind uf = new UnionFind(n);
    Set<Integer> restSet = new HashSet<>();

    for (int node : restricted)
    {
        restSet.add(node);
    }

    for (int[] edge : edges)
    {
        int a = edge[0];
        int b = edge[1];

        if (!restSet.contains(a) && !restSet.contains(b))
        {
            uf.union(a, b);
        }
    }
    return uf.getSize(0);
}

File: trees_and_graphs/reachable_nodes_with_restrictions.py
from collections import defaultdict
from collections import deque

def reachableNodesBFS(
    n: int,
    edges: list[list[int]],
    restricted: list[int]
) -> int:
    # Store all edges in neighbors
    neighbors = defaultdict(list)
    for node_a, node_b in edges:
        neighbors[node_a].append(node_b)
        neighbors[node_b].append(node_a)
    
    # Mark the nodes in restricted as visited
    seen = [False] * n
    for node in restricted:
        seen[node] = True
    
    # Store all nodes to be visited in queue
    ans = 0
    queue = deque([0])
    seen[0] = True

    while queue:
        curr_node = queue.popleft()
        ans += 1

        for next_node in neighbors[curr_node]:
            if not seen[next_node]:
                seen[next_node] = True
                queue.append(next_node)
    
    return ans

def reachableNodeDFSRecursive(
    n: int,
    edges: list[list[int]],
    restricted: list[int]
) -> int:
    neighbors = defaultdict(list)
    for node_a, node_b in edges:
        neighbors[node_a].append(node_b)
        neighbors[node_b].append(node_a)
    
    seen = [False] * n
    for node in restricted:
        seen[node] = True

    def dfsRecursive(curr_node):
        nonlocal ans
        ans += 1

        seen[curr_node] = True

        for next_node in neighbors[curr_node]:
            if not seen[next_node]:
                dfsRecursive(next_node)
    
    ans = 0
    dfsRecursive(0)
    return ans

def reachableNodeDFSIterative(
    n: int,
    edges: list[list[int]],
    restricted: list[int]
) -> int:
    neighbors = defaultdict(set)
    for node_a, node_b in edges:
        neighbors[node_a].add(node_b)
        neighbors[node_b].add(node_a)
    
    seen = [False] * n
    for node in restricted:
        seen[node] = True
    
    stack = [0]
    ans = 0
    seen[0] = True

    while stack:
        curr_node = stack.pop()
        ans += 1

        for next_node in neighbors[curr_node]:
            if not seen[next_node]:
                seen[next_node] = True
                stack.append(next_node)
    
    return ans

class UnionFind:
    def __init__(self, n):
        self.root = list(range(n))
        self.rank = [1] * n
    
    def find(self, x):
        if self.root[x] != x:
            self.root[x] = self.find(self.root[x])
        return self.root[x]
    
    def union(self, x, y):
        root_x, root_y = self.find(x), self.find(y)
        if root_x != root_y:
            if self.rank[root_x] > self.rank[root_y]:
                root_x, root_y = root_y, root_x
            
            self.rank[root_y] += self.rank[root_x]
            self.root[root_x] = root_y
    
    def getSize(self, x):
        return self.rank[self.find(x)]

def reachableNodesDSU(
    n: int,
    edges: list[list[int]],
    restricted: list[int]
) -> int:
    rest_set_ = set(restricted)
    uf = UnionFind(n)

    for node_a, node_b in edges:
        if node_a not in rest_set_ and node_b not in rest_set_:
            uf.union(node_a, node_b)
    
    return uf.getSize(0)
